### PR TITLE
Requirements update

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?

Adds spike detection for fault tolerant training

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

File: src/lightning/fabric/utilities/spike.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+import warnings
+from typing import Any, Dict, List, Literal, Optional, type_checking, Union
+
+import torch
+from lightning_utilities.core.imports import compare_version
+
+_TORCHMETRICS_GREATER_EQUAL_1_0_0 = compare_version("torchmetrics", operator.ge, "1.0.0")
+
+if type_checking:
+    from lightning.fabric.fabric import Fabric
```
Comment (awaelchli) [inline src/lightning/fabric/utilities/spike.py:13]:
```suggestion
from lightning.fabric.fabric import Fabric
```
Shouldn't be needed to add the guard, since fabric is in the package itself. 

If it was a circular import, maybe then it should be TYPE_CHECKING?

File: src/lightning/fabric/utilities/spike.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+
+class SpikeDetection:
+    """Spike Detection Callback.
+
+    Terminates training with a ``TrainingSpikeException`` when a loss-spike was detected and
+    saves the batches to skip when resuming to a file.
+
+    Currently we skip the current and the previous batch since it is unclear, whether the previous batch
+    altered the weights in a way that it causes the spike or just the current batch is corrupted somehow.
```
Comment (awaelchli) [inline src/lightning/fabric/utilities/spike.py:23]:
```suggestion
    We skip the current and the previous batch since it is unclear whether the previous batch
    altered the weights in a way that it causes the spike or just the current batch is corrupted somehow.
```
Not sure about the comma, but suggest to remove the "Currently" in the beginning to avoid frequent word repetition

File: src/lightning/fabric/utilities/spike.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+    Args:
+        mode: Whether to minimize or maximize the tracked metric
+        window: A running mean of metrics with ``window`` size. Serves as reference value for spike3s
```
Comment (awaelchli) [inline src/lightning/fabric/utilities/spike.py:27]:
```suggestion
        window: A running mean of metrics with ``window`` size. Serves as reference value for spikes.
```

File: src/lightning/fabric/utilities/spike.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        warmup: After how many batches should spike-tracking starts
```
Comment (awaelchli) [inline src/lightning/fabric/utilities/spike.py:28]:
```suggestion
        warmup: After how many batches spike-tracking should start
```

File: src/lightning/fabric/utilities/spike.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        atol: An absolute tolerance.  Every diff between the running mean and the current value,
+            that's not an improvement and above ``atol`` will be considered a spike
+        rtol: A relative tolerance. Every diff between the running mean and the current value,
+            that's higher than ``rtol * running_mean`` is considered a spike
+        exclude_batches_path: Where to save the file that contains the batches to exclude. Will default to current directory.
+        finite_only: Will consider non-finite values like NaN, inf and -inf a spike as well
```
Comment (awaelchli) [inline src/lightning/fabric/utilities/spike.py:34]: That's awesome, I like this feature!
Note that the description is the opposite of what the argument does by default. So maybe change to:
```suggestion
        finite_only: If set to ``False``, consider non-finite values like NaN, inf and -inf a spike as well.
```

File: src/lightning/fabric/utilities/spike.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+        # While spike-detection happens on a per-rank level, we need to fail all ranks if any rank detected a spike
+        is_spike_global = fabric.strategy.reduce_boolean_decision(is_spike, all=False)
+
+        if is_spike_global:
+            self.handle_spike(fabric, batch_idx)
+        else:
+            self.update_stats(loss)
+
+    def is_spike(self, loss: torch.Tensor) -> bool:
```
Comment (awaelchli) [inline src/lightning/fabric/utilities/spike.py:87]: I would make all methods protected except the hooks to call (on_train_batch_end) 2) and the state_dict methods. To clearly communicate what the user is supposed to call manually. I consider this especially important for Fabric where the user needs to call the hook.

File: tests/tests_data/conftest.py [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import pytest
```
Comment (awaelchli) [inline tests/tests_data/conftest.py:1]: probably best to delete this file in this PR

File: .github/workflows/ci-tests-data.yml [hunk 8]
```diff
@@ -97,7 +97,7 @@ jobs:
       # NOTE: do not include coverage report here, see: https://github.com/nedbat/coveragepy/issues/1003
```
Comment (awaelchli) [inline .github/workflows/ci-tests-data.yml:97]: maybe revert this file for this PR

File: tests/tests_fabric/utilities/test_spike.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+    should_raise = spike_value is None or (spike_value is not None and finite_only)
+    fabric.launch(
+        partial(
+            spike_detection_test,
+            global_rank_spike=global_rank_spike,
+            spike_value=spike_value,
+            should_raise=should_raise,
+        )
+    )
```
Comment (awaelchli) [inline tests/tests_fabric/utilities/test_spike.py:65]: Unless I misunderstand your use case, you don't need to do that. THis can be just

```suggestion
    fabric.launch(pike_detection_test,
            global_rank_spike=global_rank_spike,
            spike_value=spike_value,
            should_raise=should_raise,
    )
```
without partial. The args will be automatically forwarded to the function.

File: src/lightning/fabric/utilities/spike.py (hunk 6 above)
Comment (justusschock) [reply src/lightning/fabric/utilities/spike.py:87]: done :)

File: tests/tests_fabric/utilities/test_spike.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    for i in range(len(loss_vals)):
+        context = pytest.raises(TrainingSpikeException) if i == 4 and should_raise else contextlib.nullcontext()
+
+        with context:
+            fabric.call(
+                "on_train_batch_end",
+                fabric,
+                torch.tensor(loss_vals[i], device=fabric.device),
+                None,
+                i,
```
Comment (awaelchli) [inline tests/tests_fabric/utilities/test_spike.py:29]: consider using named args for clarity

File: tests/tests_data/conftest.py [hunk 11]
```diff
@@ -0,0 +1,10 @@
...
+import torch.distributed
+
+
+@pytest.fixture(autouse=True)
+def teardown_process_group():
```
Comment (awaelchli) [inline tests/tests_data/conftest.py:6]: I would like to understand why this is needed suddenly. Will I see an issue if I run the test without thix fixture?

File: tests/tests_fabric/utilities/test_spike.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    if fabric.global_rank == global_rank_spike:
+        if spike_value is None:
+            loss_vals[4] = 3
+        else:
+            spike_value = spike_value
+
+    for i in range(len(loss_vals)):
+        context = pytest.raises(TrainingSpikeException) if i == 4 and should_raise else contextlib.nullcontext()
+
+        with context:
```
Comment (awaelchli) [inline tests/tests_fabric/utilities/test_spike.py:23]: Not sure, but I think the problem here is that you use multi-processing but only raise the exception in one rank. So your test will probably fail. Shouldn't the the `fabric.launch()` be wrapped with a try catch too in this case?

File: src/lightning/fabric/utilities/spike.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+    """
+
+    def __init__(
+        self,
+        mode: Literal["min", "max"] = "min",
+        window: int = 10,
+        warmup: int = 1,
+        atol: Optional[float] = None,
+        rtol: Optional[float] = 2.0,
+        exclude_batches_path: Optional[str] = None,
```
Comment (awaelchli) [inline src/lightning/fabric/utilities/spike.py:44]: Consider using `utilities.imports._PATH` to support pathlib as well

File: tests/tests_fabric/utilities/test_spike.py (hunk 12 above)
Comment (justusschock) [reply tests/tests_fabric/utilities/test_spike.py:23]: no, the decision is synced in https://github.com/Lightning-AI/lightning/pull/18014/files#diff-6c07d896784502a485056fb611d5252a68c3dfa6b3eb8874444a4202750d1424R80 so both ranks raise the Exception :)

File: tests/tests_data/conftest.py (hunk 11 above)
Comment (justusschock) [reply tests/tests_data/conftest.py:6]: Otherwise I end up with not terminated processes since torch.distributed wasn't shut down. This causes tests to be a lot slower. We also have it in fabric and PL :)

File: src/lightning/fabric/utilities/spike.py [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+        finite_only: bool = True,
+    ):
+        if _TORCHMETRICS_GREATER_EQUAL_1_0_0:
+            from torchmetrics.aggregation import MeanMetric
+            from torchmetrics.wrappers import Running
+        else:
+            raise RuntimeError("SpikeDetection requires torchmetrics>=1.0.0! Please upgrade your version!")
+        super().__init__()
+
+        self.last_val: Union[torch.Tensor, float] = 0.0
```
Comment (awaelchli) [inline src/lightning/fabric/utilities/spike.py:54]: Shouldn't this be just tensor? I don't see where it could ever be just a float. Maybe standardize to just tensor to simplify the conversion in state_dict method.

**Review by awaelchli — COMMENTED** [review]
This is a great feature, thanks! Thanks for adding the integration tests. Personally I would also add a few unit tests that call into the callback directly and check some of the conditional logic.

File: tests/tests_fabric/utilities/test_spike.py (hunk 12 above)
Comment (awaelchli) [reply tests/tests_fabric/utilities/test_spike.py:23]: Okay, I missed that

File: tests/tests_data/conftest.py (hunk 11 above)
Comment (awaelchli) [reply tests/tests_data/conftest.py:6]: Oh right, but maybe delete the file here and add it in the other PR?

File: tests/tests_fabric/utilities/test_spike.py (hunk 10 above)
Comment (justusschock) [inline tests/tests_fabric/utilities/test_spike.py:28]:
```suggestion
                fabric=fabric,
                loss=torch.tensor(loss_vals[i], device=fabric.device),
                batch=None,
                batch_idx=i,
```

File: tests/tests_data/conftest.py (hunk 11 above)
Comment (justusschock) [reply tests/tests_data/conftest.py:6]: but this one uses ddp as well :) so it should be added with whichever PR lands first :)

File: src/lightning/fabric/utilities/spike.py (hunk 14 above)
Comment (justusschock) [reply src/lightning/fabric/utilities/spike.py:54]: it's a float here in the initialization because this way I don't need to move it to the correct device :D

File: tests/tests_data/conftest.py (hunk 11 above)
Comment (awaelchli) [reply tests/tests_data/conftest.py:6]: This PR only makes fabric changes. I don't see how it relates to test_data. tests for data shouldn't even run on this PR, right?

File: src/lightning/pytorch/callbacks/spike.py [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+import lightning.pytorch as pl
+from lightning.fabric.utilities.spike import SpikeDetection as FabricSpikeDetection
+from lightning.pytorch.callbacks.callback import Callback
+
+
+class SpikeDetection(FabricSpikeDetection, Callback):
+    @torch.no_grad()
+    def on_train_batch_end(  # type: ignore
+        self,
+        trainer: pl.Trainer,
```
Comment (awaelchli) [inline src/lightning/pytorch/callbacks/spike.py:15]:
```suggestion
        trainer: "pl.Trainer",
```
helps with circular imports hopefully

File: src/lightning/fabric/utilities/spike.py (hunk 13 above)
Comment (justusschock) [reply src/lightning/fabric/utilities/spike.py:44]: done

File: src/lightning/fabric/utilities/spike.py (hunk 14 above)
Comment (awaelchli) [reply src/lightning/fabric/utilities/spike.py:56]: In this case I would standardize to just float, but this is a nitpick. Feel free to resolve.

File: tests/tests_pytorch/callbacks/test_spike.py [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+            0,
+            2,
+            marks=pytest.mark.skipif(
+                sys.platform != "linux", reason="multiprocessing on other platforms takes forever"
+            ),
+        ),
+    ],
+)
+@pytest.mark.parametrize("spike_value", [None, float("inf"), float("NaN"), -float("inf")])
+@pytest.mark.parametrze("finite_only", [True, False])
```
Comment (awaelchli) [inline tests/tests_pytorch/callbacks/test_spike.py:59]:
```suggestion
@pytest.mark.parametrize("finite_only", [True, False])
```

File: tests/tests_data/conftest.py (hunk 11 above)
Comment (justusschock) [reply tests/tests_data/conftest.py:6]: oh, yeah you're right, my bad

**Review by awaelchli — APPROVED** [review]

File: src/lightning/fabric/utilities/spike.py [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+from lightning.fabric.utilities.types import _PATH
+
+
+class SpikeDetection:
+    """Spike Detection Callback.
+
+    Terminates training with a ``TrainingSpikeException`` when a loss-spike was detected and
+    saves the batches to skip when resuming to a file.
+
+    We skip the current and the previous batch since it is unclear whether the previous batch
```
Comment (lantiga) [inline src/lightning/fabric/utilities/spike.py:24]: In the future we should make this configurable so the user can decide how many batches to skip. I'm expecting this to be application-dependent.

**Review by lantiga — APPROVED** [review]
Looks great, added a comment for the future

File: requirements/fabric/base.txt [hunk 18]
```diff
@@ -7,3 +7,4 @@ fsspec[http]>2021.06.0, <2023.5.0
 packaging >=17.1, <=23.0
 typing-extensions >=4.0.0, <=4.4.0
 lightning-utilities >=0.8.0, <0.10.0
+torchmetrics >=0.7.0, <1.1.0  # needed for using fixed compare_version
```
Comment (awaelchli) [inline requirements/fabric/base.txt:10]: This should probably go into fabric/test.txt because that's the only place we need it, right?
